HPCC-30401 JTrace should optionally suppress trace/span ids generation

rpastrana opened the pull request:
- Feature request somewhat ambiguous
- Copies global.tracing to container helm values
- Generates traceid if server doesn't receive remote parent

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

rpastrana committed:
HPCC-30401 JTrace optionally suppress trace/span IDs

- Copies global.tracing to component helm values
- Revises logic to create otel span
-- Span created if per-span flag set
-- or global/component config flag set
-- or serverSpan receives valid remote parent
- Adds cppunit (cannot test scenario where no span created yet)

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexisrisk.com>

rpastrana commented:
@ghalliday it wasn't clear exactly when we wanted to force traceids. This PR is not complete but wanted to get in synch with you before spending any more time down this path...

rpastrana commented:
@ghalliday sample output from first commit doesn't make much sense. Hoping 2nd commit is closer to your intention. 

```
"!!No remoteParent received, generating remoteParent IDs: Trace '30d658bd5fbc5c6583c3dedca16b8ba0' span '208c54bca6cd4d0a' "
00000078 PRG INF 2023-10-23 13:52:02.392     7   286 UNK     "Span start: {"Type":"Server","Name":"request","GlobalID":"JD8LowRvXm3ujvmHS82NAT","LocalID":"JD8LowSwvRDqbiZ7dRDJdt","SpanID":"621030d51af8737e","TraceID":"30d658bd5fbc5c6583c3dedca16b8ba0","ParentSpanID":"208c54bca6cd4d0a"}
```

ghalliday commented on helm/hpcc/templates/thor.yaml:
```diff
@@ -32,7 +32,7 @@ Pass in dict with root and me
 {{- $eclAgentResources := .me.eclAgentResources | default dict -}}
 {{- $hthorScope := dict "name" $hthorName "jobMemorySectionName" "eclAgentMemory" | merge (pick .me "multiJobLinger" "maxGraphStartupTime") | merge (dict "resources" (deepCopy $eclAgentResources)) }}
-{{- $thorScope := omit .me "eclagent" "thoragent" "hthor" "logging" "env" "eclAgentResources" "eclAgentUseChildProcesses" "eclAgentReplicas" "thorAgentReplicas" "eclAgentType" }}
+{{- $thorScope := omit .me "eclagent" "thoragent" "hthor" "logging" "Tracing" "env" "eclAgentResources" "eclAgentUseChildProcesses" "eclAgentReplicas" "thorAgentReplicas" "eclAgentType" }}
```
should this be a lower case "tracing"?

rpastrana replied:
definitely, good catch

ghalliday commented on system/jlib/jtrace.cpp:
```diff
@@ -427,31 +427,34 @@ class CSpan : public CInterfaceOf<ISpan>
             hpccLocalId.set(ln_uid::createUniqueIdString().c_str());
 
-        auto provider = opentelemetry::trace::Provider::GetTracerProvider();
+        //we don't always want to create trace/span IDs
```
Does this preserve trace-ids if this flags are false?

rpastrana replied:
> Does this preserve trace-ids if this flags are false?

Good point, opentel remote/local parentSpans won't have a span to associate with...

rpastrana replied:
@ghalliday see my latest commit. 
If a serverspan contains a remote parent, and we're not generating trace/span IDs we'll report the remote parent Trace and span ID:

```json
{"Type":"Server","Name":"propegatedServerSpan","GlobalID":"IncomingUGID","CallerID":"IncomingCID","LocalID":"JDaYP6r3d4MJEBvgUALsZu","ParentTraceID":"beca49ca8f3138a2842e5cf21402bfff","ParentSpanID":"4b960b3e4647da3f"}
```

If we're not generating trace/spans ids, then internal/client spans could contain a local parent, but that parent would likely not have a trace/span id either, but the current commit attempts to report the local parent trace and span id anyway. I'm not sure if this is what you envisioned.

rpastrana replied:
or, the child span can silently inherit trace id from the parent

rpastrana replied:
Latest amended commit simply inherits traceID from parent when no parent otel span available:

sample output for no traceID generated mode serverSpan(propagated parent)->clientSpanserverchild->internalSpan->internalSpan2

```json
15:07:40.802583 27843 Span end: {"Type":"Internal","Name":"internalSpan2","TraceID":"beca49ca8f3138a2842e5cf21402bfff","ParentSpanID": ""}
15:07:40.802618 27843 Span end: {"Type":"Internal","Name":"internalSpan","TraceID":"beca49ca8f3138a2842e5cf21402bfff","ParentSpanID": ""}
15:07:40.802628 27843 Span end: {"Type":"Client","Name":"clientSpanserverchild","TraceID":"beca49ca8f3138a2842e5cf21402bfff","ParentSpanID": ""}
15:07:40.802637 27843 Span end: {"Type":"Server","Name":"propegatedServerSpan","GlobalID":"IncomingUGID","CallerID":"IncomingCID","LocalID":"JDbF4xnv7LSWDV4Eug1SpJ","TraceID":"beca49ca8f3138a2842e5cf21402bfff","ParentSpanID":"4b960b3e4647da3f"}
```

ghalliday commented on system/jlib/jtrace.cpp:
```diff
@@ -637,6 +638,33 @@ class CServerSpan : public CSpan
             }
         }
+
+       /*if ((!httpHeaders ||  !httpHeaders->hasProp("traceparent")) && queryTraceManager().alwaysCreateTraceIds())
```
When this is a full PR this should be moved elsewhere.

ghalliday replied:
I don't think you ever want to create a synthetic remote id.

ghalliday reviewed (commented):
@rpastrana looks good.  A couple of questions.

ghalliday commented:
I'm not sure why "Feature request somewhat ambiguous" is in the commit.  If it is ambiguous it is worth clarifying before creating the PR.

rpastrana commented:
> I'm not sure why "Feature request somewhat ambiguous" is in the commit. If it is ambiguous it is worth clarifying before creating the PR.

Because I found the description somewhat ambiguous and the creator was away and this feature was identified as priority.
Anyway, the point of that statement was for the reviewer to understand why the draft PR might have missed the gist of the feature request before significant time was spent on the implementation.

ghalliday commented on system/jlib/jtrace.cpp:
```diff
@@ -16,7 +16,7 @@
 
 
-
+//#include "opentelemetry/sdk/trace/random_id_generator_factory.h"
```
trivial: remove

ghalliday commented on system/jlib/jtrace.cpp:
```diff
@@ -185,18 +190,19 @@ class CSpan : public CInterfaceOf<ISpan>
             out.append(",\"LocalID\":\"").append(hpccLocalId.get()).append("\"");
 
+        //traceID not always tied to span object
+        out.append(",\"TraceID\":\"").append(traceID.get()).append("\"");
```
Should this check for isEmptyString()

ghalliday replied:
@rpastrana I think this question is still relevant

ghalliday commented on system/jlib/jtrace.cpp:
```diff
@@ -427,31 +436,31 @@ class CSpan : public CInterfaceOf<ISpan>
-        //what if tracerName is empty?
-        auto tracer = provider->GetTracer(tracerName.get());
+        //we don't always want to create trace/span IDs
+        if (hasMask(flags, SpanFlags::EnsureTraceId) || queryTraceManager().alwaysCreateTraceIds())
```
This doesn't seem quite right.  You still want to create a span if a trace-id was provided by the caller.  It is only omitted if it is not provided and ids are not created by default.

ghalliday commented on system/jlib/jtrace.cpp:
```diff
@@ -171,6 +171,11 @@ class CSpan : public CInterfaceOf<ISpan>
 
+    const char * getTraceID() const
+    {
+        return traceID.get();
```
This shouldn't really be stored for non-server spans, similar to caller and local spans.  The span class structure would benefit from some refactoring (but as a separate PR)

ghalliday reviewed (commented):
@rpastrana I think this is close, but it isn't quite right yet.

ghalliday reviewed (commented):
@rpastrana one question comment, but otherwise it looks ready to merge.

ghalliday reviewed (approved):
@rpastrana please squash and rebase.  Thanks.

ghalliday commented:
@rpastrana unfortunately needs rebasing again because I merged your other test.
Also, can you clarify what "disabled" means.  Does it mean don't propagate anything? How does it differ from the createXXX options?

rpastrana commented:
> @rpastrana unfortunately needs rebasing again because I merged your other test. Also, can you clarify what "disabled" means. Does it mean don't propagate anything? How does it differ from the createXXX options?

Rebased, but yes, the introduction of the create 

> @rpastrana unfortunately needs rebasing again because I merged your other test. Also, can you clarify what "disabled" means. Does it mean don't propagate anything? How does it differ from the createXXX options?

Rebased.
Yes, the newly created alwaysCreateTraceIds and its ultimate implementation does cloud the definition of the global.tracing.disabled flag...

Initially, disabled would suppress configuration of Otel's tracer(span processor/exporter, etc), and thereby avoid all otel trace/span creation/processing/exporting.
The other flags control non-otel functionality. 
As-is, the disabled and alwaysCreateTraceIds flags have an overlap.... if disabled, no valid trace IDs can be generated even if we've been instructed to "Always create TraceIds".

rpastrana commented:
@ghalliday this was showing a conflict that shouldn't have been there. After re-rebasing, I manually resolved the conflict.

ghalliday commented:
@rpastrana the commits are not quite right.  I suspect you may need to rebase your branch onto the most recent 9.4.x
